### PR TITLE
Back: Add AREA for spotify (need to talk about it with the team)

### PR DESCRIPTION
**Pull request by Beafowl-Pull**
* **Checklist before requesting a review**
- [x] The commits messages follows our guidelines
- [x] I have performed a self-review of my code
- [x] I have documented my code
- [x] Do we need to implement unit tests?

* **In what state is your pull request?**
- [x] Ready to merge / Waiting a reviwer to see my work.
- [ ] Work In Progress (WIP) / My work is not finish but i want daily reviews. (Draft)
- [ ] CI Review / I want to see what the CI thinks of my work.
- [ ] Need feedback / Just to have feedback on what i produced. (May not be merged)

* **What kind of changes does this PR introduce?** (You can choose multiple)
- [ ] Bug fix
- [x] Feature request
- [ ] New / Updated documentation
- [ ] Testing CI ( Make the pull request in draft mode)

## Summary by CodeRabbit

- **New Features**
  - Introduced a new Spotify integration that tracks changes in user's liked songs, albums, artists, and playlists.

- **Improvements**
  - Enhanced Spotify service to handle OAuth token retrieval and API calls more efficiently.
  - Streamlined Spotify triggers for detecting new saved content.

- **Bug Fixes**
  - Fixed an issue where Spotify actions weren't properly handling errors.

- **Refactor**
  - Updated function signatures for clarity and consistency across Spotify services.

- **Chores**
  - Performed a database reset to ensure clean synchronization with the new service structure.

### CODE REVIEW

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts [hunk 1]
```diff
@@ -2,22 +2,32 @@
...
+                headers: {
+                    "Authorization": "Bearer " + oauthToken
+                }
+            });
+            const json: any = await response.json();
+            console.log(json);
+        } catch (e) {
+            console.error(e);
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (Beafowl-Pull) [reply back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:30]: it make 'throw' of exception caught locally no ?

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts [hunk 2]
```diff
@@ -29,9 +39,24 @@ namespace SpotifyTriggers {
...
         return true;
     }
```
[the remark this answers is not in the document]
Comment (Beafowl-Pull) [reply back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:62]: how do i use this function after ?

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts [hunk 3]
```diff
@@ -2,22 +2,74 @@
...
+            const json = await fetchWithOAuth(oauthId, ownerId, SPOTIFY_API_BASE_URL + "/me/tracks");
+            const userData = getOrCreateUserData(ownerId);
+            userData.SoundLikedLength = json.items.length;
+            console.log(json);
+        } catch (e) {
+            console.error("Error in checkSpotifyNewSavedSong:", e);
+            return false;
+        }
         return true;
     }
```
[the remark this answers is not in the document]
Comment (Beafowl-Pull) [reply back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:74]: wdym by that ?

File: back/src/services/servicesApps/reactions/spotify.reactions.servicesApp.ts [hunk 4]
```diff
@@ -19,16 +19,26 @@ namespace SpotifyReactions {
...
-        if (oauthToken === null) {
-            console.error("OAuth token not found");
+        try {
+            const oauthToken: string | null = await OAuthService.getDecryptedOAuthTokenFromId(oauthId, ownerId);
+            if (oauthToken === null) {
+                console.error("OAuth token not found");
+                return false;
+            }
+            const actionDataParsed: any = actionData;
+            const reactionDataParsed: any = reactionData;
```
Comment (romainpanno) [inline back/src/services/servicesApps/reactions/spotify.reactions.servicesApp.ts:29]: We're in ts, do not put ts

File: back/src/services/servicesApps/reactions/spotify.reactions.servicesApp.ts [hunk 5]
```diff
@@ -19,16 +19,26 @@ namespace SpotifyReactions {
...
+            const response: any = await fetch("https://api.spotify.com/v1/users/" + actionDataParsed.userId + "/playlists/" + reactionDataParsed.playlistId + "/tracks?uris=" + actionDataParsed.trackUri, {
```
Comment (romainpanno) [inline back/src/services/servicesApps/reactions/spotify.reactions.servicesApp.ts:30]: Check if these elements are in the json object

File: back/src/services/servicesApps/reactions/spotify.reactions.servicesApp.ts [hunk 6]
```diff
@@ -19,16 +19,26 @@ namespace SpotifyReactions {
...
+                method: "POST",
+                headers: {
+                    "Authorization": "Bearer " + oauthToken
+                }
+            });
+            const json: any = await response.json();
```
Comment (romainpanno) [inline back/src/services/servicesApps/reactions/spotify.reactions.servicesApp.ts:36]: Type

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts [hunk 7]
```diff
@@ -2,22 +2,74 @@
...
+ * @constant {Record<string, SpotifyTriggerData>} usersSpotifyTriggerData - The data of the Spotify triggers
+ */
+const SPOTIFY_API_BASE_URL = "https://api.spotify.com/v1";
+const usersSpotifyTriggerData: Record<string, SpotifyTriggerData> = {};
+
+/**
+ * @interface SpotifyTriggerData
+ * @description The data of a Spotify trigger
+ */
+interface SpotifyTriggerData {
```
Comment (romainpanno) [inline back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:18]: Create this interface in the interfaces folder

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts [hunk 8]
```diff
@@ -2,22 +2,74 @@
...
  */
 namespace SpotifyTriggers {
-    /**
-     * Check if a new song has been saved
-     * @param ownerId - The owner id of the trigger
-     * @param oauthId - The oauth id of the trigger
-     * @param data - The data of the trigger
-     * @returns {Promise<boolean>} - The result of the trigger
-     */
+    async function fetchWithOAuth(oauthId: number, ownerId: number, url: string): Promise<any> {
```
Comment (romainpanno) [inline back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:31]: Add doc

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts [hunk 9]
```diff
@@ -2,22 +2,74 @@
...
+    }
+
     export const checkSpotifyNewSavedSong = async (ownerId: number, oauthId: number, data: JSON): Promise<boolean> => {
-        console.log(ownerId);
-        console.log(oauthId);
-        console.log(data);
+        try {
+            const json = await fetchWithOAuth(oauthId, ownerId, SPOTIFY_API_BASE_URL + "/me/tracks");
+            const userData = getOrCreateUserData(ownerId);
+            userData.SoundLikedLength = json.items.length;
```
Comment (romainpanno) [inline back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:67]: Does this modify the reference ?

Comment (romainpanno) [reply back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:67]: Check if lenght is different and return false if not

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts [hunk 10]
```diff
@@ -29,9 +81,15 @@ namespace SpotifyTriggers {
      * @returns {Promise<boolean>} - The result of the trigger
      */
     export const checkSpotifyNewSavedAlbum = async (ownerId: number, oauthId: number, data: JSON): Promise<boolean> => {
-        console.log(ownerId);
-        console.log(oauthId);
-        console.log(data);
+        try {
+            const json = await fetchWithOAuth(oauthId, ownerId, SPOTIFY_API_BASE_URL + "/me/albums");
+            const userData = getOrCreateUserData(ownerId);
+            userData.AlbumLikedLength = json.items.length;
```
Comment (romainpanno) [inline back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:87]: Same

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts [hunk 11]
```diff
@@ -43,9 +101,15 @@ namespace SpotifyTriggers {
      * @returns {Promise<boolean>} - The result of the trigger
      */
     export const checkSpotifyNewSavedArtist = async (ownerId: number, oauthId: number, data: JSON): Promise<boolean> => {
-        console.log(ownerId);
-        console.log(oauthId);
-        console.log(data);
+        try {
+            const json = await fetchWithOAuth(oauthId, ownerId, SPOTIFY_API_BASE_URL + "/me/following?type=artist");
+            const userData = getOrCreateUserData(ownerId);
+            userData.ArtistLikedLength = json.artists.items.length;
```
Comment (romainpanno) [inline back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:107]: Same

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts [hunk 12]
```diff
@@ -57,9 +121,15 @@ namespace SpotifyTriggers {
      * @returns {Promise<boolean>} - The result of the trigger
      */
     export const checkSpotifyNewSavedPlaylist = async (ownerId: number, oauthId: number, data: JSON): Promise<boolean> => {
-        console.log(ownerId);
-        console.log(oauthId);
-        console.log(data);
+        try {
+            const json = await fetchWithOAuth(oauthId, ownerId, SPOTIFY_API_BASE_URL + "/me/playlists");
+            const userData = getOrCreateUserData(ownerId);
+            userData.PlaylistLikedLength = json.items.length;
```
Comment (romainpanno) [inline back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:127]: Same

**Review by romainpanno — CHANGES_REQUESTED** [review]
Add types and change any in good type

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts (hunk 8 above)
Comment (Beafowl-Pull) [reply back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:31]: oupsiiii

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts (hunk 9 above)
Comment (Beafowl-Pull) [reply back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:67]: yes it is according with chat gpt

File: back/src/interfaces/spotify.interface.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+ * @interface SpotifyTriggerData
+ * @description The data of a Spotify trigger
+ */
+interface SpotifyTriggerData {
+    userId: string;
+    TrackLikedLength: number;
+    AlbumLikedLength: number;
+    ArtistLikedLength: number;
+    PlaylistLikedLength: number;
+}
```
Comment (romainpanno) [inline back/src/interfaces/spotify.interface.ts:15]: EOF

File: back/src/types/action.type.ts [hunk 14]
```diff
@@ -0,0 +1,4 @@
+type ActionType = {
+    userId: string;
+    trackUri: string;
+}
```
Comment (romainpanno) [inline back/src/types/action.type.ts:4]: EOF

File: back/src/types/reaction.type.ts [hunk 15]
```diff
@@ -0,0 +1,3 @@
+type ReactionType = {
+    playlistId: string;
+}
```
Comment (romainpanno) [inline back/src/types/reaction.type.ts:3]: EOF

**Review by romainpanno — CHANGES_REQUESTED** [review]

**Review by romainpanno — APPROVED** [review]

**Review by KitetsuK — APPROVED** [review]

**Pull request merged**